runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactStagingServiceTest.java

```diff
@@ -251,17 +258,129 @@ public void putArtifactsMultipleFilesTest() throws Exception {
     assertFiles(files.keySet(), stagingToken);
   }
 
+  @Test
+  public void putArtifactsMultipleFilesConcurrentlyTest() throws Exception {
+    String stagingSession = "123";
+    Map<String, Integer> files = new HashMap<>();
+    files.put("file5cb", (DATA_1KB / 2) /*500b*/);
+    files.put("file1kb", DATA_1KB /*1 kb*/);
+    files.put("file15cb", (DATA_1KB * 3) / 2  /*1.5 kb*/);
+    files.put("nested/file1kb", DATA_1KB /*1 kb*/);
+    files.put("file10kb", 10 * DATA_1KB /*10 kb*/);
+    files.put("file100kb", 100 * DATA_1KB /*100 kb*/);
+
+    final String text = "abcdefghinklmop\n";
+    files.forEach((fileName, size) -> {
+      Path filePath = Paths.get(srcDir.toString(), fileName).toAbsolutePath();
+      try {
+        Files.createDirectories(filePath.getParent());
+        Files.write(filePath,
+            Strings.repeat(text, Double.valueOf(Math.ceil(size * 1.0 / text.length())).intValue())
+                .getBytes(CHARSET));
+      } catch (IOException ignored) {
+      }
+    });
+    String stagingSessionToken = BeamFileSystemArtifactStagingService
+        .generateStagingSessionToken(stagingSession, destDir.toUri().getPath());
+
+    List<ArtifactMetadata> metadata = new ArrayList<>();
+    ExecutorService executorService = Executors.newFixedThreadPool(8);
+    try {
+      for (String fileName : files.keySet()) {
+        executorService.execute(() -> {
+          try {
+            putArtifact(stagingSessionToken,
+                Paths.get(srcDir.toString(), fileName).toAbsolutePath().toString(), fileName);
+          } catch (Exception e) {
+            Assert.fail(e.getMessage());
+          }
+          metadata.add(ArtifactMetadata.newBuilder().setName(fileName).build());
+        });
+      }
+    } finally {
+      executorService.shutdown();
+      executorService.awaitTermination(2, TimeUnit.SECONDS);
+    }
+
+    String stagingToken = commitManifest(stagingSessionToken, metadata);
+    Assert.assertEquals(
+        Paths.get(destDir.toAbsolutePath().toString(), stagingSession, "MANIFEST").toString(),
+        stagingToken);
+    assertFiles(files.keySet(), stagingToken);
+  }
+
+  @Test
+  public void putArtifactsMultipleFilesConcurrentSessionsTest() throws Exception {
+    String stagingSession1 = "123";
+    String stagingSession2 = "abc";
+    Map<String, Integer> files = new HashMap<>();
```

Comment:
Nit (here and above): I prefer using ImmutableMaps for constants like this.

Reply:
ImmutableMap.of(k,v) is only applicable for 5 k-v while we are having more KVs.
Using builders to create the immutable map.